### PR TITLE
Aura Spells have internal Attach Spell for multiple Enchant Keywords

### LINKED ISSUE #169: Refactor Aura Spells + Enchant Keyword
Rules wise Aura Spells should just check for their Enchant Keywords and these should handle the "can be attached" part.

* The Corner cases: having an Aura become the copy of `Old-Growth Troll`, when it dies it returns as real Aura and it is gaining `enchant Forest` . now with two Enchant keywords, both needs to be respected.
* if that's cleaned up, the Aura Spell can be created by CardFactory like other Spell Permanent are.
* Enchant Targets are: Cards on the Battlefield, Cards in Graveyard, Players
* The AI Logic can't be part of the Aura Spell anymore and should probably be either part of the keyword, or be part of an extra SVar.
* cleanup other uses for FirstAttachSpell

* The cleaned up enchant keyword needs to respect `Tallowisp` that explicit looks for `enchant creature`, not `enchant creature you control`

Rules:
> 15.1b Aura spells are always targeted. An Aura’s target is specified by its enchant keyword ability (see rule 702.5, “Enchant”)
> 702.5c If an Aura has multiple instances of enchant, all of them apply. The Aura’s target must follow the restrictions from all the instances of enchant. The Aura can enchant only objects or players that match all of its enchant abilities.

An Aura without enchant keyword would be unattached too. (and sent to the graveyard)

Ruling for _Immovable Rod_
> If you target an Aura with Immovable Rod's last ability, it will lose its enchant ability and be put into its owner's graveyard because it can't legally enchant anything. An Equipment that loses its abilities loses its equip ability, so you won't be able to move it that way, but it will stay attached to the creature it's equipping, if any. (2021-07-23)

### PR DESCRIPTION
**Pull request by Hanmac**
Closes #169 

I can't add the AI flags into the Keyword because of `Tallowisp`,
so i added Svars instead.

- [x] Updated Cards
- [x] Fixed GUI, i think?
- [ ] AI still need to use filter the possible list via `canBeAttached`

### CODE REVIEW

**Comment by Hanmac** [comment]
@tool4ever i don't know why " Incremental Growth " fails ... that isn't even an Aura?

**Comment by Hanmac** [comment]
@Agetian can you look at the failing test case why the AI has problem with Incremental Growth?
i somehow can't see how this is related to this MR

**Comment by Agetian** [comment]
Yeah, agree about the AI logic improvements, sounds like a good thing to implement :+1:

File: forge-game/src/main/java/forge/game/card/CardFactory.java [hunk 1]
```diff
@@ -443,7 +443,37 @@ private static void readCardFace(Card c, ICardFace face) {
...
+                if (c.hasSVar("AttachAITgts")) {
+                    extra += " | AITgts$ " + c.getSVar("AttachAITgts");
+                }
+                if (c.hasSVar("AttachAILogic")) {
+                    extra += " | AILogic$ " + c.getSVar("AttachAILogic");
+                }
+                if (c.hasSVar("AttachAIValid")) { // TODO combine with AttachAITgts
+                    extra += " | AIValid$ " + c.getSVar("AttachAIValid");
+                }
+                String st = "SP$ Attach | ValidTgts$ Card,Player | TgtZone$ Battlefield,Graveyard | TgtPrompt$ Select target " + desc + extra;
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/card/CardFactory.java:471]: I don't understand how this is supposed to work
where do you restrict the targeting?

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/card/CardFactory.java:471]: Via `GameEntity.canBeEnchantedBy` where i check all Enchant Keywords (for corner cases)

**Comment by Hanmac** [comment]
tests are green, and i did some hard wire that `canTarget` does call `canAttach` for Aura Spells

**Comment by tool4ever** [comment]
yea that's what I meant
but wouldn't adding `CanBeEnchantedBy` property to the ValidTgts$ be much cleaner? :thinking:

**Comment by Hanmac** [comment]
> yea that's what I meant but wouldn't adding `CanBeEnchantedBy` property to the ValidTgts$ be much cleaner? 🤔

I don't know? Like this? `ValidTgts$ Card.CanBeEnchantedBy,Player.CanBeEnchantedBy`

You can try it if you want?

**Comment by Hanmac** [comment]
@tool4ever can you check why the last change does break the tests again?

`getAllCandidates` there calls `sa.canTarget` which then should call `tr.getValidTgts()` ?

**Comment by Hanmac** [comment]
@tool4ever the MR should be clean now?
@Agetian now that `CanBeEnchantedBy` is part of the `ValidTgts`, the AI should respect that?

**Comment by Agetian** [comment]
> @Agetian now that `CanBeEnchantedBy` is part of the `ValidTgts`, the AI should respect that?

Hopefully yes, but if you can tell me some test cases for that, I can give it a shot and test/trace it in action :)
Otherwise, looks good, yeah :+1:

**Comment by Hanmac** [comment]
@Agetian it is mostly the problem that `tgt.canTgtPlayer()` doesn't work anymore
but i think the one place in AttachAi where that is used got updated

File: forge-ai/src/main/java/forge/ai/ability/AttachAi.java [hunk 2]
```diff
@@ -551,11 +552,7 @@ private static Card attachAIReanimatePreference(final SpellAbility sa, final Lis
         final Player ai = sa.getActivatingPlayer();
         final Card attachSourceLki = CardCopyService.getLKICopy(attachSource);
         attachSourceLki.setLastKnownZone(ai.getZone(ZoneType.Battlefield));
```
[the remark this answers is not in the document]
Comment (Hanmac) [reply forge-ai/src/main/java/forge/ai/ability/AttachAi.java:554]: It got tested by `GameSimulationTest.testEpochrasite`

Or should we do another Test case for this?

File: forge-ai/src/main/java/forge/ai/ability/AttachAi.java [hunk 3]
```diff
@@ -940,9 +941,28 @@ protected boolean doTriggerAINoCost(final Player ai, final SpellAbility sa, fina
...
+                if (v.contains("Player") || v.contains("Opponent")) {
+                    spellCanTargetPlayer = true;
+                    break;
+                }
+            }
+        }
+        if (tgt.canTgtPlayer() && (!isAuraSpell(sa) || spellCanTargetPlayer)) {
             List<Player> targetable = new ArrayList<>();
             for (final Player player : sa.getHostCard().getGame().getPlayers()) {
+                if (isAuraSpell(sa) && !player.canBeAttached(sa.getHostCard(), sa)) {
```
Comment (tool4ever) [inline forge-ai/src/main/java/forge/ai/ability/AttachAi.java:963]: at least this part should no longer be needed?

**Comment by Agetian** [comment]
I agree that it's important to come up with some in-game board states that would be good to check how this works in practice, it's a bit difficult for me to figure out what would be good test cases tbh, but yeah, the automated tests won't help since they use the simulation AI which is a completely different thing in itself :)

**Comment by Agetian** [comment]
Hmm yeah, not sure what's happening there, probably needs looking into, yes. I'll see if I can figure it out on the upcoming weekend.

File: forge-ai/src/main/java/forge/ai/ability/AttachAi.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hanmac) [reply forge-ai/src/main/java/forge/ai/ability/AttachAi.java:554]: @tool4ever you or @Agetian could try to add an extra `Animate Dead` Test that uses AI parts

[the remark this answers is not in the document]
Comment (tool4ever) [reply forge-ai/src/main/java/forge/ai/ability/AttachAi.java:554]: I don't think we have a precedent for that...? So we'd need to add some framework parts there first.
Anyway to fix the logic here I suggest you try using `AnimateAi.becomeAnimated` so the new Aura is able to pass the check

[the remark this answers is not in the document]
Comment (Hanmac) [reply forge-ai/src/main/java/forge/ai/ability/AttachAi.java:554]: @tool4ever i will need to look at Weekend how to change the new Aura

**Comment by Hanmac** [comment]
Some things that might be done before this MR:

- `KeepTapped` for Prison Spells could probably more abstracted so that the AttachAi logic can look for the CantHappen Replacement Effect, then no extra AiLogic needed.
- I want to add `THISTYPE` similar to `CARDNAME` so it does create `this aura` in the description. (for this i need to look how it works for different languages)

File: forge-ai/src/main/java/forge/ai/ability/AttachAi.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hanmac) [reply forge-ai/src/main/java/forge/ai/ability/AttachAi.java:596]: @tool4ever added AnimateAi logic

[the remark this answers is not in the document]
Comment (tool4ever) [reply forge-ai/src/main/java/forge/ai/ability/AttachAi.java:596]: ok it works now

**Review by tool4ever — APPROVED** [review]
I don't know about the final TODO but otherwise it's big enough :p
